data/decoder-builder.js

```diff
@@ -0,0 +1,292 @@
+        { from: 0x94, to: 0x201D },
+        { from: 0x95, to: 0x2022 },
+        { from: 0x96, to: 0x2013 },
+        { from: 0x97, to: 0x2014 },
+        { from: 0x98, to: 0x02DC },
+        { from: 0x99, to: 0x2122 },
+        { from: 0x9A, to: 0x0161 },
+        { from: 0x9B, to: 0x203A },
+        { from: 0x9C, to: 0x0153 },
+        { from: 0x9E, to: 0x017E },
+        { from: 0x9F, to: 0x0178 }
+    ]
+    return `
+            j = 1;
+            var cc = seg.charCodeAt(1);
+            var num = 0;
+            var isEmpty;
+            if ((cc == ${'x'.charCodeAt(0)}) || (cc == ${'X'.charCodeAt(0)})) {
+                do {
+                    cc = seg.charCodeAt(++j);
+                    if ((cc > ${'0'.charCodeAt(0)-1}) && (cc < ${'9'.charCodeAt(0)+1})) { num = num * 16 + cc - ${'0'.charCodeAt(0)}; }
+                    else if ((cc > ${'a'.charCodeAt(0)-1}) && (cc < ${'f'.charCodeAt(0)+1})) { num = num * 16 + cc - ${'a'.charCodeAt(0) - 10}; }
+                    else if ((cc > ${'A'.charCodeAt(0)-1}) && (cc < ${'F'.charCodeAt(0)+1})) { num = num * 16 + cc - ${'A'.charCodeAt(0) - 10}; }
+                    else break;
+                } while (1)
+                isEmpty = j <= 2;
+            } else {
+                while (1) {
+                    if ((cc < ${'0'.charCodeAt(0)}) || (cc > ${'9'.charCodeAt(0)})) break;
+                    num = num * 10 + cc - ${'0'.charCodeAt(0)};
+                    cc = seg.charCodeAt(++j);
+                }
+                isEmpty = j < 1;
+            }
+            if (isEmpty) {
+                parseError("${ERRORS.MISSING_DIGIT.MSG}",${ERRORS.MISSING_DIGIT.CODE});
+                output += '&' + seg;
+                continue;
+            }
+            if (cc == ${';'.charCodeAt(0)}) {
+                j++;
+            } else if (strict) {
+                parseError("${ERRORS.MISSING_SEMICOLON.MSG}",${ERRORS.MISSING_SEMICOLON.CODE});
+                output += '&' + seg;
+                continue;
+            }
+            if (num > ${0x10FFFF}) {
+                parseError("${ERRORS.OUT_OF_RANGE_CHAR_REF.MSG}",${ERRORS.OUT_OF_RANGE_CHAR_REF.CODE});
+                output += '\\uFFFD' + seg.substring(j);
+            } else if (num == 0) {
+                parseError("${ERRORS.NULL_CHAR_REF.MSG}",${ERRORS.NULL_CHAR_REF.CODE});
+                output += '\\uFFFD' + seg.substring(j);
+            } else if ( (num > ${0xD800-1}) && (num < ${0xDFFF+1}) ) {
+                parseError("${ERRORS.SURROGATE_CHAR_REF.MSG}",${ERRORS.SURROGATE_CHAR_REF.CODE});
+                output += '\\uFFFD' + seg.substring(j);
+            } else {` + /* https://infra.spec.whatwg.org/#c0-control */ `
+                if (((num > ${0xFDD0-1}) && (num < ${0xFDEF+1})) || ([${NON_CHARACTER.join(',')}].indexOf(num) >= 0)) {
+                    parseError("${ERRORS.NON_CHARACTER.MSG}",${ERRORS.NON_CHARACTER.CODE});
+                } else if ((num == ${0x0D}) || (num < ${0x001F+1})) {
```

Comment:
Mistake here. C0 control characters other than \x09 (tab), \x0A (lf), \x0C (ff), and \x20 (space) are invalid. Should be
`} else if ( (num < ${0x20}) && (num != ${0x09}) && (num != ${0x0A}) && (num != ${0x0C}) ) { `